drivers/gpu/virtio/gpu.c

```diff
@@ -0,0 +1,767 @@
+    regs->QueueDeviceLow = (virtio_metadata_paddr + used_off) & 0xFFFFFFFF;
+    regs->QueueDeviceHigh = (virtio_metadata_paddr + used_off) >> 32;
+    regs->QueueReady = 1;
+
+    /* Finish initialisation */
+    regs->Status |= VIRTIO_DEVICE_STATUS_DRIVER_OK;
+}
+
+static gpu_resp_status_t virtio_gpu_to_sddf_resp_status(enum virtio_gpu_ctrl_type type)
+{
+    switch (type) {
+    case VIRTIO_GPU_RESP_OK_DISPLAY_INFO:
+    case VIRTIO_GPU_RESP_OK_NODATA:
+        return GPU_RESP_OK;
+    case VIRTIO_GPU_RESP_ERR_INVALID_RESOURCE_ID:
+        return GPU_RESP_ERR_INVALID_RESOURCE_ID;
+    case VIRTIO_GPU_RESP_ERR_INVALID_SCANOUT_ID:
+        return GPU_RESP_ERR_INVALID_SCANOUT_ID;
+    case VIRTIO_GPU_RESP_ERR_INVALID_PARAMETER:
+        return GPU_RESP_ERR_INVALID_PARAMETER;
+    case VIRTIO_GPU_RESP_ERR_OUT_OF_MEMORY:
+    case VIRTIO_GPU_RESP_ERR_INVALID_CONTEXT_ID:
+    case VIRTIO_GPU_RESP_ERR_UNSPEC:
+        return GPU_RESP_ERR_UNSPEC;
+    default:
+        LOG_GPU_VIRTIO_DRIVER_ERR("Unexpected virtIO response type received from device: 0x%x\n", type);
+        assert(false);
+        return GPU_RESP_ERR_UNSPEC;
+    }
+}
+
+static enum virtio_gpu_formats sddf_to_virtio_gpu_resource_format(gpu_formats_t format)
+{
+    switch (format) {
+    case GPU_FORMAT_B8G8R8A8_UNORM:
+        return VIRTIO_GPU_FORMAT_B8G8R8A8_UNORM;
+    case GPU_FORMAT_B8G8R8X8_UNORM:
+        return VIRTIO_GPU_FORMAT_B8G8R8X8_UNORM;
+    case GPU_FORMAT_A8R8G8B8_UNORM:
+        return VIRTIO_GPU_FORMAT_A8R8G8B8_UNORM;
+    case GPU_FORMAT_X8R8G8B8_UNORM:
+        return VIRTIO_GPU_FORMAT_X8R8G8B8_UNORM;
+    case GPU_FORMAT_R8G8B8A8_UNORM:
+        return VIRTIO_GPU_FORMAT_R8G8B8A8_UNORM;
+    case GPU_FORMAT_R8G8B8X8_UNORM:
+        return VIRTIO_GPU_FORMAT_R8G8B8X8_UNORM;
+    case GPU_FORMAT_A8B8G8R8_UNORM:
+        return VIRTIO_GPU_FORMAT_A8B8G8R8_UNORM;
+    case GPU_FORMAT_X8B8G8R8_UNORM:
+        return VIRTIO_GPU_FORMAT_X8B8G8R8_UNORM;
+    default:
+        LOG_GPU_VIRTIO_DRIVER_ERR("Unexpected sDDF resource format: 0x%x\n", format);
+        assert(false);
+        return VIRTIO_GPU_FORMAT_B8G8R8A8_UNORM;
+    }
+}
+
+static bool handle_response()
+{
+    int err = 0;
+    bool sddf_notify = false;
+    uint16_t i = last_handled_used_idx;
+    uint16_t curr_idx = virtq.used->idx;
+    while (i != curr_idx) {
+        struct virtq_used_elem used = virtq.used->ring[i % virtq.num];
+        assert(used.id < VIRTQ_QUEUE_SIZE);
+
+        LOG_GPU_VIRTIO_DRIVER("Handling response %d\n", virtio_desc_to_id[used.id]);
+        
+        struct virtq_desc desc_head = virtq.desc[used.id];        
+        assert(desc_head.len >= sizeof(struct virtio_gpu_ctrl_hdr));
+        assert(desc_head.flags & VIRTQ_DESC_F_NEXT);
+        assert(desc_head.next < VIRTQ_QUEUE_SIZE);
+
+        err = ialloc_free(&ialloc_desc, used.id);
+        assert(!err);
+        err = ialloc_free(&ialloc_desc, desc_head.next);
+        assert(!err);
```

Comment:
Does this deallocate all the resources?  Because further down, you refer also to `desc[desc_head.next].next` in the `footer` for `CREATE_RESOURCE_BLOB`

Reply:
The order doesn't matter here assuming a non-multithreaded PD, because reallocation will only happens after this handler function. But yes it will break in concurrent environments and I agree it does make for a confusing read. Will be moving this down.